File: react/rasa/mcp_client.py
import asyncio
import websockets
import json
import os
from typing import Dict, Any, Optional

class MCPClient:
    def __init__(self, server_url: str, api_key: str = None, api_secret: str = None):
        self.server_url = server_url
        self.api_key = api_key
        self.api_secret = api_secret
        self.websocket = None
        self.request_id = 1

    async def connect(self):
        """Connect to the MCP server"""
        try:
            # Try different authentication methods
            headers = {}
            if self.api_key and self.api_secret:
                headers['Authorization'] = f'Bearer {self.api_key}:{self.api_secret}'

            self.websocket = await websockets.connect(
                self.server_url,
                extra_headers=headers
            )
            print("✅ Connected to MCP server")
            return True
        except Exception as e:
            print(f"❌ Failed to connect to MCP server: {e}")
            return False

    async def disconnect(self):
        """Disconnect from the MCP server"""
        if self.websocket:
            await self.websocket.close()
            self.websocket = None

    async def query(self, method: str, params: Dict[str, Any]) -> Optional[Dict[str, Any]]:
        """Send a query to the MCP server"""
        if not self.websocket:
            print("❌ Not connected to MCP server")
            return None

        try:
            request = {
                "jsonrpc": "2.0",
                "id": self.request_id,
                "method": method,
                "params": params
            }

            self.request_id += 1

            await self.websocket.send(json.dumps(request))
            print(f"📤 Sent: {json.dumps(request, indent=2)}")

            response = await self.websocket.recv()
            result = json.loads(response)
            print(f"📥 Received: {json.dumps(result, indent=2)}")

            return result

        except Exception as e:
            print(f"❌ Query failed: {e}")
            return None

# Example usage
async def test_mcp_connection():
    client = MCPClient(
        server_url="ws://localhost:3000",
        api_key="mdb_sa_id_68baeeb922d3970b63bba9e7",
        api_secret="mdb_sa_sk_VneO3QlphelgmBIm4oULVYiZOx5lo8ObPFpVHNQF"
    )

    if await client.connect():
        # Test find query
        result = await client.query("find", {
            "database": "mcp_database",
            "collection": "employees",
            "filter": {},
            "limit": 5
        })

        if result:
            print("✅ Query successful!")
            print(f"Result: {result}")
        else:
            print("❌ Query failed")

        await client.disconnect()
    else:
        print("❌ Could not connect to MCP server")

if __name__ == "__main__":
    asyncio.run(test_mcp_connection())

File: Chat/backend/mcp_client.js
const WebSocket = require('ws');

class McpClient {
  constructor(options) {
    this.serverUrl = options.serverUrl;
    this.apiKey = options.apiKey;
    this.apiSecret = options.apiSecret;
    this.ws = null;
    this.requestId = 1;
    this.pendingRequests = new Map();
  }

  async connect() {
    return new Promise((resolve, reject) => {
      console.log(`Connecting to MCP server at ${this.serverUrl}...`);

      this.ws = new WebSocket(this.serverUrl, {
        headers: {
          'Authorization': `Bearer ${this.apiKey}:${this.apiSecret}`,
          'X-API-Key': this.apiKey,
          'X-API-Secret': this.apiSecret,
          'Content-Type': 'application/json'
        }
      });

      this.ws.on('open', () => {
        console.log('✅ Connected to MCP server');
        resolve();
      });

      this.ws.on('message', (data) => {
        try {
          const response = JSON.parse(data.toString());
          const requestId = response.id;

          if (this.pendingRequests.has(requestId)) {
            const { resolve, reject } = this.pendingRequests.get(requestId);
            this.pendingRequests.delete(requestId);

            if (response.error) {
              reject(new Error(response.error.message));
            } else {
              resolve(response.result);
            }
          }
        } catch (error) {
          console.error('Error parsing MCP response:', error);
        }
      });

      this.ws.on('error', (error) => {
        console.error('WebSocket error:', error);
        reject(error);
      });

      this.ws.on('close', (code, reason) => {
        console.log(`Connection closed: ${code} - ${reason}`);
      });

      // Timeout after 10 seconds
      setTimeout(() => {
        reject(new Error('Connection timeout'));
      }, 10000);
    });
  }

  async query(method, params = {}) {
    if (!this.ws || this.ws.readyState !== WebSocket.OPEN) {
      throw new Error('Not connected to MCP server');
    }

    const request = {
      jsonrpc: '2.0',
      id: this.requestId++,
      method: method,
      params: params
    };

    return new Promise((resolve, reject) => {
      this.pendingRequests.set(request.id, { resolve, reject });
      this.ws.send(JSON.stringify(request));

      // Timeout after 30 seconds
      setTimeout(() => {
        if (this.pendingRequests.has(request.id)) {
          this.pendingRequests.delete(request.id);
          reject(new Error('Request timeout'));
        }
      }, 30000);
    });
  }

  async find(collection, options = {}) {
    const params = {
      database: 'mcp_database',
      collection: collection,
      filter: options.filter || {},
      limit: options.limit || 10
    };

    if (options.projection) {
      params.projection = options.projection;
    }

    return this.query('find', params);
  }

  async insert(collection, documents) {
    const params = {
      database: 'mcp_database',
      collection: collection,
      documents: Array.isArray(documents) ? documents : [documents]
    };

    return this.query('insert', params);
  }

  async update(collection, filter, update, options = {}) {
    const params = {
      database: 'mcp_database',
      collection: collection,
      filter: filter,
      update: update,
      upsert: options.upsert || false
    };

    return this.query('update', params);
  }

  async delete(collection, filter) {
    const params = {
      database: 'mcp_database',
      collection: collection,
      filter: filter
    };

    return this.query('delete', params);
  }

  disconnect() {
    if (this.ws) {
      this.ws.close();
      this.ws = null;
    }
  }
}

module.exports = McpClient;
